### PR TITLE
healthmap.org crawler

### PR DESCRIPTION
**Pull request by aishwaryakapse**
This crawler runs by entering the number of days to crawl as a command line parameter.

#74 
@JavierJia @chenlica @MoniMoledo @kaiyim

### CODE REVIEW

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+# create the parameters, needs start_date and end_date in "mm/dd/yyyy" format
+def generate_health_map_query(start_date, end_date):
+    parameter_map = {"category":["1","2","29"],"diseases":["282"], "heatscore":"1", "partner":"hm"}
+    parameter_map['sdate'] = start_date
+    parameter_map['edate'] = end_date
+    return parameter_map
+    
+    
+def getUrl(alertid, trto, trfr, pid):
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: could you rename the `trto` and `trfr` to meaningful names?

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    query['trfr'] = trfr
+    url = "http://www.healthmap.org/ai.php?"+alertid+"&"+urllib.urlencode(query)+"&"+"pid"+pid
+    return url
+
+def storeHtml(response, fname, headers):
+    request = urllib2.Request(url, headers = headers)
+    response = urllib2.urlopen(request)
+    with io.open(fname+".html", "w", encoding="utf-8") as writeFile:
+        writeFile.write(response.read().decode("utf-8"))
+        writeFile.write(u"\u000A")
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: why do we need to write the `000A`?

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+if __name__ == '__main__':
+    
+    #headers
+    headers = {}
+    headers['User-Agent'] = "Mozilla/5.0 (X11; Linux i686) AppleWebKit/537.17 (KHTML, like Gecko) Chrome/24.0.1312.27 Safari/537.17"
+    
+    dirName = sys.argv[1]
+    fileName = sys.argv[2]
+    days = sys.argv[3]
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: can you add some argument check? and print out the usage once if fails.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    #get the end date and start date of 1 days interval
+    end_date = date_counter.strftime("%m/%d/%Y")
+    start_date = (date_counter - interval).strftime("%m/%d/%Y")
+       
+    while end_date != condition:
+        newDirName = dirName+"\\"+str(date_counter)
+        if not os.path.exists(newDirName):
+            os.makedirs(newDirName)
+            
+        completeName = os.path.join(newDirName, fileName)
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: nit: any function larger than 10 lines should be further broken down to modular functions.
e.g. the logic is something like
```
while end_date != condition:
     resultFile = storeJsonRecord(curDate)
     storeHTMLFrom(resultFile)
     increaseDate
```
and those large chunks could be put into smaller function. Just an idea, you can choose your own function names.

File: noah/src/main/resources/crawler/pahoconf.conf [hunk 5]
```diff
@@ -5,5 +5,7 @@ url1: http://www.paho.org/hq/index.php?option=com_content&view=article&id=11959&
 
 url2: http://ais.paho.org/phip/viz/ed_zika_cases.asp
 
+url3: http://www.paho.org/hq/index.php?option=com_content&view=article&id=11599&Itemid=41691&lang=en
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/pahoconf.conf]: why do we need to change this file?

**Comment by JavierJia** [comment]
looks good in general, i left some minor comments

File: noah/src/main/resources/crawler/health_map_crawler.py (hunk 1 above)
Comment (aishwaryakapse) [inline noah/src/main/resources/crawler/health_map_crawler.py]: Lol. Ok. They are like that on the website. I don't know what they mean exactly. They are query parameters. I will try to give them meaningful names.

Comment (aishwaryakapse) [inline noah/src/main/resources/crawler/health_map_crawler.py]: Done.

File: noah/src/main/resources/crawler/health_map_crawler.py (hunk 2 above)
Comment (aishwaryakapse) [inline noah/src/main/resources/crawler/health_map_crawler.py]: That is the unicode character \n i.e; new line. It is added at the end after writing the whole file. I don't think it is needed. I deleted it.

Done.

File: noah/src/main/resources/crawler/health_map_crawler.py (hunk 3 above)
Comment (aishwaryakapse) [inline noah/src/main/resources/crawler/health_map_crawler.py]: Yes.. Added this thing.

Done.

File: noah/src/main/resources/crawler/pahoconf.conf (hunk 5 above)
Comment (aishwaryakapse) [inline noah/src/main/resources/crawler/pahoconf.conf]: Need to discuss about the crawler for paho.org.

File: noah/src/main/resources/crawler/health_map_crawler.py (hunk 4 above)
Comment (aishwaryakapse) [inline noah/src/main/resources/crawler/health_map_crawler.py]: Got it. thanks for this advice. The code looks very clean now.

Done.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import urllib
+import datetime
+import io
+import os
+import json
+import sys
+
+def argumentCheck():
+    if len(sys.argv) != 4:
+        sys.exit("Invalid number of Arguments... \n Enter the directory name, filename where json data needs to be stored and number of days in arguments: \n Example: python health_map_crawler.py HealthMapResult 365")
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: the example is wrong? 
`python health_map_crawler.py HealthMapResult 365` only has 2 arguments.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    if len(sys.argv) != 4:
+        sys.exit("Invalid number of Arguments... \n Enter the directory name, filename where json data needs to be stored and number of days in arguments: \n Example: python health_map_crawler.py HealthMapResult 365")  
+    dirName = sys.argv[1]
+    fileName = sys.argv[2]
+    days = sys.argv[3]
+    directoryCheck(dirName)
+    return dirName, fileName, days
+
+def getDatesAndCondition():
+    # get today's date
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: useless comment.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    dateCounter = datetime.date.today()
+    #get the stop condition date
+    condition = (dateCounter - datetime.timedelta(int(days))).strftime("%m/%d/%Y")
+    # set interval to 1 day
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: code tells. don't write unnecessary comments.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    interval = datetime.timedelta(1)
+    return dateCounter, condition, interval
+
+def getStartAndEndDate(dateCounter, interval):
+    #get the end date and start date of 1 days interval
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: comment is wrong. this function can take any interval.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    endDate = dateCounter.strftime("%m/%d/%Y")
+    startDate = (dateCounter - interval).strftime("%m/%d/%Y")
+    return endDate, startDate 
+     
+#get health map response object
+def getHealthMapResponse(url, headers):
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: the code is exactly the same as the comment

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    request = urllib2.Request(url, headers = headers)
+    response = urllib2.urlopen(request)
+    return response
+
+# create the parameters, needs start_date and end_date in "mm/dd/yyyy" format
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: this is a good one.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    query['trfr'] = queryParamFr
+    url = "http://www.healthmap.org/ai.php?"+alertid+"&"+urllib.urlencode(query)+"&"+"pid"+pid
+    return url
+    
+def storeJsonResult(completeName, startDate, endDate, headers):
+    with io.open(completeName + ".json", "w", encoding="utf-8") as writeFile:
+            # get response from healthmap
+            url = "http://www.healthmap.org/getAlerts.php?"
+            response = getHealthMapResponse(url + urllib.urlencode(generateHealthMapQuery(startDate, endDate)), headers)
+            #write to file
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: useless comments.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+def getEachUrl(alertid, queryParamTo, queryParamFr, pid):
+    query = {}
+    query['trto'] = queryParamTo
+    query['trfr'] = queryParamFr
+    url = "http://www.healthmap.org/ai.php?"+alertid+"&"+urllib.urlencode(query)+"&"+"pid"+pid
+    return url
+    
+def storeJsonResult(completeName, startDate, endDate, headers):
+    with io.open(completeName + ".json", "w", encoding="utf-8") as writeFile:
+            # get response from healthmap
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: code tells.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            (alertid, queryParamTo, queryParamFr, pid) = getQueryParam(html)
+            url = getEachUrl(alertid, queryParamTo, queryParamFr, pid)
+            response = getHealthMapResponse(url, headers)
+            fname = os.path.join(newDirName, alertid)           
+            request = urllib2.Request(url, headers = headers)
+            response = urllib2.urlopen(request)
+            with io.open(fname+".html", "w", encoding="utf-8") as writeFile:
+                writeFile.write(response.read().decode("utf-8"))
+        
+def getQueryParam(html):
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: this one deserves some comments.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    end = html.find(")", start)
+    result = html[start:end]
+    elements = result.split(',')
+    alertid = elements[0]
+    queryParamTo = elements[1].strip('\'')
+    queryParamFr = elements[2].strip('\'')
+    pid = elements[3]
+    return alertid, queryParamTo, queryParamFr, pid
+    
+def directoryCheck(dirName):
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: it is not only a `check`. it actually create the dir. should be something like `createDirIfNotExists`.

File: noah/src/main/resources/crawler/health_map_crawler.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    if not os.path.exists(dirName):
+        os.makedirs(dirName)
+    
+if __name__ == '__main__':
+    #headers
```
Comment (JavierJia) [inline noah/src/main/resources/crawler/health_map_crawler.py]: clear all those unnecessary comments.

**Comment by JavierJia** [comment]
code is good in general. there is one saying `good code don't need comments`. use correct variable/method names to explain the purpose of the code. 

The good comments should be only put on top the APIs (in your case, all those functions) to explain the contents and the argument protocol and the return values. comments in between code usually is distracting.

**Comment by JavierJia** [comment]
👍

**Pull request merged**